### PR TITLE
vinyl: fix bugs in cache that can lead to existing tuples missing in select results

### LINKED ISSUE #7943: Pagination does not work correctly with vinyl_cache
Let's run the script:
```lua
s.execute('rm -rf *.snap *.xlog *.vylog 512')

json = require('json')
box.cfg{}

s = box.schema.create_space('vy_test', {engine='vinyl'})
s:create_index'pk'
for i = 1, 10 do s:replace{i, 0} end

s:select(nil, {fullscan=true, after={5}})
print(json.encode(s:select(nil, {fullscan=true, after={1}})))
```

Actual output:
[[6,0],[7,0],[8,0],[9,0],[10,0]]
Expected output:
[[2,0],[3,0],[4,0],[5,0],[6,0],[7,0],[8,0],[9,0],[10,0]]

The problem is not observed when vinyl cache is disabled.

### LINKED ISSUE #7947: `select()` from a vinyl space may skip a tuple if the tuple cache is enabled
Reproduced with Tarantool 2.11.0-entrypoint-698-g215630e6f6c4. Tarantool 1.10 and 2.10 are affected, as well.

Reproducer:

```Lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

box.cfg{log_level = 4}

box.schema.create_space('test', {engine = 'vinyl'})
box.space.test:create_index('pk', {parts = {1, 'unsigned', 2, 'unsigned'}})
box.space.test:insert{1, 1}
box.space.test:insert{1, 2}
box.space.test:insert{1, 3}

box.begin()
box.space.test:delete{1, 2}
box.space.test:select{1}
box.rollback()

print("select{1}")
for _, t in box.space.test:pairs{1} do
    print(t)
end
print("select{1, 2}")
for _, t in box.space.test:pairs{1, 2} do
    print(t)
end

os.exit()
```

Output:

```Lua
select{1}
[1, 1]
[1, 3]
select{1, 2}
[1, 2]
```

Disabling the tuple cache (`box.cfg{vinyl_cache = 0}`) makes the issue disappear.

### PR DESCRIPTION
**Pull request by locker**
Closes #7943
Closes #7947

### CODE REVIEW

File: src/box/vy_cache.c [hunk 1]
```diff
@@ -226,9 +226,11 @@ vy_cache_env_set_quota(struct vy_cache_env *env, size_t quota)
 
 void
 vy_cache_add(struct vy_cache *cache, struct vy_entry curr,
-	     struct vy_entry prev, struct vy_entry key,
-	     enum iterator_type order)
+	     struct vy_entry prev, bool is_boundary,
```
Comment (drewdzzz) [inline src/box/vy_cache.c:229]: Nit: `is_boundary` is a stale name - the flag no longer indicates whether the tuple is the final tuple in the chain - only the starting one.

File: test/engine-luatest/pagination_test.lua [hunk 2]
```diff
@@ -557,6 +557,19 @@ tree_g.test_empty_page_with_offset = function(cg)
     end)
```
Comment (drewdzzz) [inline test/engine-luatest/pagination_test.lua:557]: Please, enable the cache for this test suite.

**Review by drewdzzz — CHANGES_REQUESTED** [review]

File: test/engine-luatest/pagination_test.lua (hunk 2 above)
Comment (locker) [reply test/engine-luatest/pagination_test.lua:557]: Well, the cache's enabled by default, but you're right - better enable it explicitly. Done it like this:

<details><summary>diff</summary>

```diff
diff --git a/test/engine-luatest/pagination_test.lua b/test/engine-luatest/pagination_test.lua
index 17eaac125d4a..c295a1df0579 100644
--- a/test/engine-luatest/pagination_test.lua
+++ b/test/engine-luatest/pagination_test.lua
@@ -557,6 +557,21 @@ tree_g.test_empty_page_with_offset = function(cg)
     end)
 end
 
+tree_g.before_test('test_gh_7943', function(cg)
+    -- Vinyl tuple cache must be enabled to reproduce the issue.
+    cg.server:exec(function()
+        rawset(_G, 'vinyl_cache', box.cfg.vinyl_cache)
+        box.cfg({vinyl_cache = 1024 * 1024})
+    end)
+end)
+
+tree_g.after_test('test_gh_7943', function(cg)
+    cg.server:exec(function()
+        box.cfg({vinyl_cache = rawget(_G, 'vinyl_cache')})
+        rawset(_G, 'vinyl_cache', nil)
+    end)
+end)
+
 tree_g.test_gh_7943 = function(cg)
     cg.server:exec(function()
         local t = require('luatest')
```
</details>

File: src/box/vy_cache.c (hunk 1 above)
Comment (locker) [reply src/box/vy_cache.c]: Fair enough. Renamed `vy_cache_add::is_boundary` and `vy_read_iterator::is_cache_boundary` to `vy_cache_add::is_first` and `vy_read_iterator::is_first_cached`, respectively

**Review by drewdzzz — APPROVED** [review]

**Pull request merged**